src/cargo/sources/git/utils.rs

```diff
@@ -880,53 +883,157 @@ pub fn fetch(
+        return fetch_with_cli(repo, orig_url, &refspecs, tags, config);
     }
+    if config
+        .cli_unstable()
+        .gitoxide
+        .map_or(false, |git| git.fetch)
+    {
+        let git2_repo = repo;
+        let config_overrides = cargo_config_to_gitoxide_overrides(config)?;
+        let repo_reinitialized = AtomicBool::default();
+        let res = oxide::with_retry_and_progress(
+            &git2_repo.path().to_owned(),
+            config,
+            &|repo_path,
+              should_interrupt,
+              mut progress,
+              url_for_authentication: &mut dyn FnMut(&gix::bstr::BStr)| {
+                // The `fetch` operation here may fail spuriously due to a corrupt
+                // repository. It could also fail, however, for a whole slew of other
+                // reasons (aka network related reasons). We want Cargo to automatically
+                // recover from corrupt repositories, but we don't want Cargo to stomp
+                // over other legitimate errors.
+                //
+                // Consequently we save off the error of the `fetch` operation and if it
+                // looks like a "corrupt repo" error then we blow away the repo and try
+                // again. If it looks like any other kind of error, or if we've already
+                // blown away the repository, then we want to return the error as-is.
+                loop {
+                    let res = oxide::open_repo(
+                        repo_path,
+                        config_overrides.clone(),
+                        oxide::OpenMode::ForFetch,
+                    )
+                    .map_err(crate::sources::git::fetch::Error::from)
+                    .and_then(|repo| {
+                        debug!("initiating fetch of {:?} from {}", refspecs, orig_url);
+                        let url_for_authentication = &mut *url_for_authentication;
+                        let remote = repo
+                            .remote_at(orig_url)?
+                            .with_fetch_tags(if tags {
+                                gix::remote::fetch::Tags::All
+                            } else {
+                                gix::remote::fetch::Tags::Included
+                            })
+                            .with_refspecs(
+                                refspecs.iter().map(|s| s.as_str()),
+                                gix::remote::Direction::Fetch,
+                            )
+                            .map_err(crate::sources::git::fetch::Error::Other)?;
+                        let url = remote
+                            .url(gix::remote::Direction::Fetch)
+                            .expect("set at init")
+                            .to_owned();
+                        let connection =
+                            remote.connect(gix::remote::Direction::Fetch, &mut progress)?;
+                        let mut authenticate = connection.configured_credentials(url)?;
```

Comment:
[`with_authentication`](https://github.com/rust-lang/cargo/blob/0c331721d91a154de2ef8f50ffab572e829fb0cb/src/cargo/sources/git/utils.rs#L477) prioritize SSH over `credential.helper`. Maybe I understand it wrongly but it seems that the same rule doesn't hold in [gix default authentication callback](https://github.com/Byron/gitoxide/blob/ad23f1dd58b3eabf5a12d02c4a1de8b1835051d5/gix/src/config/snapshot/credential_helpers.rs#L47).

In addition, I have no idea if it works but [`git::Cred::default()`](https://docs.rs/git2/0.16/git2/struct.Cred.html#method.default) seems to respect NTLM and Kerberos. Do we have that via the default credential helper?

It's totally reasonable to follow what git does, yet this is undoubtedly a breaking change if the old rule doesn't hold. I do feel when we stabilize gioxide feature in Cargo, we can advertise it follows what git does and ditch custom rules.

Reply:
Indeed, by nature these implementations will be different, even though I hope ultimately they have the same result. Due to a lack of tests (due to the incredible difficulty of setting them up) it's not easy to say what behaviour would actually occour and I am not able to run these test cases by hand either.

Due to `gitoxide` being by default like git but flexible enough to probably be like `git2` as well, I believe we can get close to what `git2` does once it's a bit clearer what that is.

> In addition, I have no idea if it works but [git::Cred::default()](https://docs.rs/git2/0.16/git2/struct.Cred.html#method.default) seems to respect NTLM and Kerberos. Do we have that via the default credential helper?

That depends on the credentials helper that is ultimately invoked, but I don't know. It seems these are specifically added by `git2` and I don't know what it takes to add them to `gitoxide` - it mostly delegates to the credentials helper and that seems like the right thing to do.

In any case, I think it's worth keeping an eye on this topic and get to the trial stage to have users report the cases they need to work, and then we make them work as needed. If ultimately that means to setup the credential helper correctly, that *seems* like a fair thing to do as it's something most people have to do once for git anyway.